### ISSUE TITLE
Could you support MSE loss?

### ISSUE THREAD

**Issue by liaocs2008** [StanfordVL/MinkowskiEngine#27]
Hi,

MSE loss in pytorch cannot be directly applied since your sparse tensor shape utilizes the feature vector's shape. Subtraction might cause dimension mismatching or unexpected broadcasting. 

Could you help support this feature? Thanks!

**Comment by chrischoy** [comment]
I don't understand the question.

Could you read how the quantization of the space and the labels are defined? https://stanfordvl.github.io/MinkowskiEngine/utils.html#MinkowskiEngine.utils.sparse_quantize

**Comment by liaocs2008** [comment]
Thanks for your prompt reply!

I tested indoor.py with sparse_quantize function and think it is mainly for continuous coordinates where you might need to quantize (group) several points together. 

I am trying to do something like image super-resolution, where I need to calculate the MSE loss of two sparse images. Here is what I did:

> \# image x in shape [10, 32, 32], image y in shape [2, 32, 32]
xcoords = np.nonzero(x)
xfeat = x[coords]
new_xcoords = np.array(xcoords).T
sx = ME.SparseTensor(torch.from_numpy(xfeat), coords=torch.from_numpy(new_xcoords))

For calculating the MSE loss, the image y has to be a torch float tensor. (I tried representing y with your sparse tensor and found the error none "require_grad")

> criterion = torch.nn.MSELoss()
ty = torch.from_numpy(y)
output = net(sx)
loss = criterion(output.F, ty)
loss.backward()

Since output here is a sparse tensor, its shape is different from ty. 

Could you give me some suggestion how to solve it?

**Comment by chrischoy** [comment]
Unfortunately, this is not as straight-forward as the regular dense tensors. If you want, you can use the `dense()` function (https://github.com/StanfordVL/MinkowskiEngine/blob/master/MinkowskiEngine/SparseTensor.py#L394).

However, you have to use the indices generated from the voxelization process to align the labels with the input. For example, https://github.com/StanfordVL/MinkowskiEngine/blob/master/examples/two_dim_training.py#L130 creates labels that you can directly associate with the input tensor.

However, to create a higher resolution (smaller tensor stride) tensor, you have to get the coordinate using `output.C` and extract labels at the correct locations to create your required label. Once you do that, you can use the MSELoss.

This is a rather complicated procedure, but I cannot think of a better way to generate a label right now.

**Comment by liaocs2008** [comment]
I was thinking you would provide support for subtraction between your sparse tensors or sparse tensor and dense tensor. 

Thank you though!

**Comment by chrischoy** [comment]
I see.

If either of the input is dense, it would be okay to use the provided `.to_dense()` to convert the sparse tensor to a dense one.
